### ISSUE TITLE
Slow training with Vehicle loss functions

### ISSUE THREAD

**Issue by wenkokke** [vehicle-lang/vehicle#432]
@ndslusarz Could you provide some details? What test reproduces the behaviour, what test establishes the baseline.

**Comment by ndslusarz** [comment]
This problem occurs when running `vehicle-python/tests/test_mnist` in the branch `loss_experiments`.

1. For a baseline - running standard-mnist-training which simply trains the exact copy of the model but with standard python and keras implementation - it takes roughly 2-3 minutes on my laptop.

2. Running the `test_mnist` file wthout any custom vehicle loss (merely cross entropy but passed through the training code) - takes rougly 15 min per epoch (I'm saying roughly as it varies depending on whether I was working on anything else but it was the quickest time I observed. 

3. Running test_mnist with a custom loss function (specifically calling DL2 translation) gets even slower - around 40ish minutes per epoch.

One possible way I came up with to make 3 slightly faster may be by impementing finite quantifiers (forallIn) properly so that they do not have to sample and merely have a finite for loop. This would possibly help with speed on constraints involving iterating over indices of vectors or output labels.

Aside from this I'm unsure if this is something that can be improved and is merely slowed down by an unoptimal implementation or if it's simply this slow due to us using custom loss and not the optimised keras versions (1). There is a good change speed of 3 can be improved as that is dependant on the translation implementation (in `vehicle-python/vehicle/__init__`) but I have no idea by how much.

**Comment by wenkokke** [comment]
@ndslusarz Do the loss functions returned by `vehicle.generate_loss_function` work with [`tf.function`]?

[`tf.function`]: https://www.tensorflow.org/api_docs/python/tf/function
[`tf.Graph`]: https://www.tensorflow.org/api_docs/python/tf/Graph
[`tf.autograph`]: https://www.tensorflow.org/api_docs/python/tf/autograph

**Comment by ndslusarz** [comment]
I'm not sure, I'll have to check tomorrow. Unless @MatthewDaggitt or @Tgl70 one of you knows off the top of your head?

Since I updated for the other issue:

2. the code for running this setup this can now be found in `vehicle-python/tests/test_mnist_cross_entropy.py`

3. the code for running this setup this is now in `vehicle-python/tests/test_mnist_custom_loss.py`

**Comment by ndslusarz** [comment]
Marco doesn't believe it can work with tf.function. However we have managed to speed up the training to mere minutes . Mainly by increasing the size of batches - we had some issues with quantifiers interfering with batches before but it seems after we fixed the previous issue with quantifiers probably a week ago it is not causing errors anymore. I will be taking a look tomorrow into whether the larger batches work fine with the quantifier implementation in theoretical sense and don't break anything. If it does not cause any issues it seems to solve the slow training problem (it's slower than cookie-cutter python and keras but takes less than 5-10 minutes for all translations which I am alright with).

So I believe the problem is sloved - even if I do find an error somewhere it will be with quantifier implementation, not the speed. I'll hold off with closing the issue until I'm sure though.

**Comment by MatthewDaggitt** [comment]
We have entirely reworked training so I'm going to close this.
